transforms/stage-4-new-years/index.js

```diff
@@ -0,0 +1,176 @@
+const path = require('node:path');
+
+const { nameMapFile } = require('../../common/constants');
+const { getPackageRoot } = require('../../common/filesystem');
+const { readJsonFile } = require('../../common/json');
+const { toKebabCase } = require('../../common/naming');
+
+function buildResolver(b, name, type) {
+  return b.sexpr('resolve', [b.string(name)], b.hash([b.pair('type', b.string(type))]));
+}
+
+function getUnknownNames(namesByAddon) {
+  const unknownNames = Object.create(null);
+
+  for (const [type, names] of Object.entries(namesByAddon.unknown)) {
+    for (const name of names) {
+      if (!(name in unknownNames)) {
+        unknownNames[name] = [];
+      }
+
+      unknownNames[name].push(type);
+    }
+  }
+
+  return unknownNames;
+}
+
+function isDefinedByLet(node, nodePath) {
+  let current = nodePath;
+
+  while (current) {
+    if (
+      current.node.type === 'BlockStatement' &&
+      current.node.path.original === 'let' &&
+      current.node.program.blockParams.includes(node.tag)
+    ) {
+      return true;
+    }
+
+    current = current.parent;
+  }
+
+  return false;
+}
+
+module.exports = function ({ source, path: templatePath }, { parse, visit }) {
+  const absTemplatePath = path.resolve(templatePath);
+  const packageRoot = getPackageRoot(absTemplatePath);
+  const pathFromRoot = path.relative(packageRoot, absTemplatePath);
+  const namesMapPath = path.join(packageRoot, nameMapFile);
+  const namesMap = readJsonFile(namesMapPath);
+
+  if (!(pathFromRoot in namesMap)) {
+    console.error(
+      `Skipping '${templatePath}' because it could not be found in its package's names map` +
+        ` ('${namesMapPath}'). Make sure stages 2 and 3 completed successfully.`
+    );
+
+    return;
+  }
```

Comment:
Instead of doing a `console.error()` here (and, if I missed it, similarly in earlier stages), let's *return* the errors to the caller. That way the caller can choose how to handle them:

- by just dumping them all to the console immediately *itself*, in some modes
- by writing a JSON/YAML/etc. file with a well-structured object to work with
- by generating a Markdown file which someone can read and work through
- etc.

I would return some kind of structured error object—it could even be a `new Error()`, but it doesn't really need to be. Something like this:

```js
return {
  ok: false,
  message: string,
};
```